packages/expo-file-system/ios/FilePickingUtils.swift

```diff
@@ -0,0 +1,131 @@
+import ExpoModulesCore
+import MobileCoreServices
+import UIKit
+import UniformTypeIdentifiers
+
+internal protocol FilePickingResultHandler {
+  func didPickFileAt(url: URL)
+  func didCancelPicking()
+}
+
+internal struct FilePickingContext {
+  let promise: Promise
+  let initialUri: URL?
+  let mimeType: String?
+  let isDirectory: Bool
+  let delegate: FilePickingDelegate
+  var pickedUrl: URL?
+}
+
+internal class FilePickingDelegate: NSObject, UIDocumentPickerDelegate, UIAdaptivePresentationControllerDelegate {
+  private let resultHandler: FilePickingResultHandler
+  private let isDirectory: Bool
+  private weak var pickingHandler: FilePickingHandler?
+
+  init(resultHandler: FilePickingResultHandler, isDirectory: Bool = false) {
+    self.resultHandler = resultHandler
+    self.isDirectory = isDirectory
+    self.pickingHandler = resultHandler as? FilePickingHandler
+  }
+
+  func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
+    guard let url = urls.first else {
+      self.resultHandler.didCancelPicking()
+      return
+    }
+
+    self.resultHandler.didPickFileAt(url: url)
```

Comment:
should we support picking several files/directories? It's now a good time to add this to API type (return FileSystemFile/ FileSystemFile[])